pkg/sosreport/index.js

```diff
@@ -0,0 +1,180 @@
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+define([
+    "jquery",
+    "base1/cockpit",
+    "translated!base1/po"
+], function($, cockpit, po) {
+    "use strict";
+
+    cockpit.locale(po);
+    cockpit.translate();
+    var _ = cockpit.gettext;
+
+    var sos_task;
+    var sos_archive_url;
+    var sos_archive_files;
+    var sos_pid;
+
+    function sos_init() {
+        // Start right away
+        sos_create();
+    }
+
+    function sos_error(message, extra) {
+        $("#sos-alert, #sos-progress, #sos-download").hide();
+        $("#sos-error .alert-message").text(message);
+        if (extra) {
+            $("#sos-error-extra").text(extra);
+            $("#sos-error-extra").show();
+        } else
+            $("#sos-error-extra").hide();
+        $("#sos-error").show();
+        $("#sos-cancel").text(_("Close"));
+    }
+
+    function sos_create() {
+        $("#sos-progress .progress-bar").css("width", "0%");
+        $("#sos-download, #sos-error").hide();
+        $("#sos-alert, #sos-progress").show();
+        $("#sos-cancel").text(_("Cancel"));
+
+        sos_archive_url = null;
+        sos_archive_files = [ ];
+        sos_pid = null;
+
+        var task = cockpit.script('echo PID:$$; /usr/sbin/sosreport --batch', [ ],
+                                  { superuser: true, err: "out" });
+        sos_task = task;
+
+        // XXX - use a real API instead of scraping stdout
+        var output = "";
+        var pid_regex = /^PID:([0-9]+)\n/;
+        var progress_regex = /Running ([0-9]+)\/([0-9]+):/g;
+        var archive_regex = /Your sosreport has been generated and saved in:[ \n]+(\/[^\n]+)/;
+
+        task.stream(function (text) {
+            if (sos_task == task) {
+                var m, p;
+
+                output += text;
+
+                if (!sos_pid && (m = pid_regex.exec(output))) {
+                    sos_pid = m[1];
+                    output = output.slice(m[0].length);
+                }
+
+                p = 0;
+                while ((m = progress_regex.exec(output))) {
+                    p = (parseInt(m[1]) / parseInt(m[2])) * 100;
+                }
+                $("#sos-progress .progress-bar").css("width", p.toString()+"%");
+            }
+        });
+        task.done(function () {
+            if (sos_task == task) {
+                var m = archive_regex.exec(output);
+                if (m) {
+                    var archive = m[1];
+                    var basename = archive.replace(/.*\//, "");
+
+                    sos_archive_files = [ archive, archive + ".md5" ];
+
+                    var query = window.btoa(JSON.stringify({
+                        payload: "fsread1",
+                        binary: "raw",
+                        path: archive,
+                        superuser: true,
+                        external: {
+                            "content-disposition": 'attachement; filename="' + basename + '"',
```

Comment:
Does the basename need escaping or encoding? What can a valid Content-Disposition header file name include?